worker/controlsocket/worker.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controlsocket
+
+import (
+	"net"
+	"net/http"
+	"runtime"
+
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/catacomb"
+
+	"github.com/juju/juju/juju/sockets"
+)
+
+// logger is here to stop the desire of creating a package level logger.
+// Don't do this, instead use the one passed as manifold config.
+type logger interface{}
+
+var _ logger = struct{}{}
+
+// Logger represents the methods used by the worker to log information.
+type Logger interface {
+	Debugf(string, ...interface{})
+	Warningf(string, ...interface{})
+	Infof(string, ...interface{})
+}
+
+// Config represents configuration for the controlsocket worker.
+type Config struct {
+	state      State
+	logger     Logger
+	socketName string
+}
+
+// Validate returns an error if config cannot drive the Worker.
+func (config Config) Validate() error {
+	if config.state == nil {
+		return errors.NotValidf("nil state")
+	}
+	if config.logger == nil {
+		return errors.NotValidf("nil logger")
+	}
+	if config.socketName == "nil" {
+		return errors.NotValidf("empty socketName")
+	}
+	return nil
+}
+
+// Worker is a controlsocket worker.
+type Worker struct {
+	Config
+	catacomb catacomb.Catacomb
```

Comment:
This should be a tomb.

Reply:
why's that?

Reply:
Because a tomb manages 1 worker's lifecycle, a catacomb manages many workers. It makes the worker easier to understand, as you'll never need to reason about, whether it manages many or one.

Also, we're coming to the decision that there should be a series base workers that you can compose from, that will give you the right thing from the start.

Maybe something akin to:

```sh
type SingularHandler interface {
	Run(ctx context.Context) error
}

type SingularWorker struct {
	tomb tomb.Tomb
	handler Handle
}

func NewSingularWorker(handler Handle) *SingularWorker {
	w := &SingularWorker{
		handler: handler,
	}
	w.tomb.Go(func() error {
		ctx, cancel := context.WithCancel(w.tomb.Context(context.Background()))
		go func() {
			defer cancel()
			select {
				case <-w.tomb.Dying():
				case <-ctx.Done():
			}
		}
		return w.handler.Run(ctx)
	})
	return w
}

func (w *SingularWorker) Kill() {
	w.tomb.Kill(nil)
}

func (w *SingularWorker) Wait() error {
	return w.tomb.Wait()
}
```

Reply:
yeah it would be really nice to have some base workers or utilities, there is currently a **lot** of boilerplate in writing a worker